Userland/Libraries/LibWeb/HTML/HTMLFormElement.cpp

```diff
@@ -1012,4 +1038,59 @@ WebIDL::ExceptionOr<JS::Value> HTMLFormElement::named_item_value(FlyString const
     return node;
 }
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#default-button
+JS::GCPtr<HTMLButtonElement> HTMLFormElement::default_button()
+{
+    // A form element's default button is the first submit button in tree order whose form owner is that form element.
+    for (auto element : m_associated_elements) {
```

Comment:
I'm not sure whether m_associated_elements are in tree order?

One could contrive a bonkers example like this that might screw up this logic:

```html
<div id="divvy">
	<button id="first" type="submit" form="defaultButton" onclick='console.log("button1")'></button>
	<form id="defaultButton">
	    <input type="submit" />
	    <input type="text"/>
	</form>
</div>
<script>
     let div = document.getElementById("divvy");
     let btn = document.createElement("button");
     btn.id = "second";
     btn.type = "submit";
     btn.setAttribute('form', defaultButton.id);
     console.log(defaultButton.id);
     btn.addEventListener('click', (e) => { e.preventDefault(); console.log("button2") });	
     div.insertBefore(btn, div.children[0]); 
</script>
```

Reply:
Looks like there's 3 bugs here :sweat_smile: 
1. tree order
2. That `<button id=first>` element doesn't appear at all in `m_associated_elements`, so I'm guessing we're not handling `form="defaultButton"` correctly.
3. The `defaultButton` impl here isn't considering `<input type=submit>` either.

I've pushed a fix and test case for 3 while I look at 2.

Reply:
Oh whoops, I knew about a bug here. See the named and/or indexed properties form tests for proof that form attribute works if the element is *after* the form in tree order.

Reply:
I pushed a fix for 2 (generally to handle `form` attributes changing, and element `id` attributes changes, to reset the form-associated elements' owner).

For tree order, I guess we just need to traverse the tree? Otherwise, not sure how to keep `m_associated_elements` sorted in tree order

Reply:
Yep, exactly. in both [elements()](https://html.spec.whatwg.org/multipage/forms.html#dom-form-elements) and [named_item_value()](https://html.spec.whatwg.org/multipage/forms.html#dom-form-nameditem) We have to traverse the tree starting at the form element's root() and check "are you a form associated element whose form is me"?

Otherwise, we'll have to start actually sorting m_associated_elements according to tree order, which is... very uncomfortable. We'd have to come up with some more interesting ways to store that information that gets updated when the tree is updated that doesn't tank performance. (any more than it already sucks to traverse the while dang tree looking for form elements :sweat_smile: )

Reply:
Alrighty, this should be fixed. Added tests for the default button being outside the `<form>` tree (both before and after the form), and a test for dynamically inserting a button before an existing button.